File: groupanagrams.py
from collections import defaultdict
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        ans = defaultdict(list)
        for string in strs:
            sorted_str = sorted(string)
            ans[tuple(sorted_str)].append(string)
        
        return list(ans.values())
        

File: Group Anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        
        if(strs == null || strs.length == 0) return null;
        HashMap<String, List<String>> map = new HashMap<>();

        for(String current: strs){//n
            char[] characters = current.toCharArray();
            Arrays.sort(characters);//klogk
            //String sorted = new String(characters);
            String sortedStr = String.valueOf(characters);
            if(!map.containsKey(sortedStr)){
                map.put(sortedStr, new ArrayList<>());//k
            }
            map.get(sortedStr).add(current);
        }
        
        return new ArrayList<>(map.values());
    }
}